torch/_torch_docs.py

```diff
@@ -3491,6 +3491,32 @@ def merge_dicts(*dicts):
             [ 3,  3]])
 """)
 
+add_docstr(torch.nonzero_tuple,
```

Comment:
I wonder if it is better to add a kwarg to `nonzero`.

Reply:
@SsnL I don't think it makes sense to have a kwarg controlling the output type (tuple vs. tensor)

Reply:
Hmm, but we have several existing examples: torch.lu, torch.unique, torch.unique_consecutive

Reply:
I agree with @SsnL here -- otherwise we are going to have non-numpy behavior forever for no good reason.

How about something like this:
- add a new parameter (return_tuple?) defaulting to None.
- None for now means old behavior, you can turn on a deprecation warning on None, saying None will change to mean True in the next release, if they want old behavior, pass False, otherwise pass True.
- In the next release, change the default.

Reply:
If `nonzero` default behavior is okayed to change, #1834 is also related. NumPy returns a squeezed tensor (differently from PyTorch): 
```python
np.random.rand(10).nonzero()[0].shape
# (10,)
```

Reply:
@vadimkantorov this PR does exactly that, and follows numpy semantics

Reply:
It seems there is a consensus here, so I will make this a kwarg.

Reply:
It seems the way we are supporting multiple possible return types (torch.lu , torch.unique, etc.) is by returning a tuple of (tensor, tensor, tensor) in C++, then extracting values as necessary in Python:

```python
def unique_consecutive(input, return_inverse=False, return_counts=False, dim=None):
    # snip
    output, inverse_indices, counts = torch._C._VariableFunctions.unique_consecutive(
        input, return_inverse=return_inverse, return_counts=return_counts, dim=dim)
    if return_inverse and return_counts:
        return output, inverse_indices, counts
    if return_inverse:
        return output, inverse_indices
    if return_counts:
        return output, counts
    return output
```

So, the logical way to implement this would be to make `nonzero` return a `std::tuple<Tensor, TensorList>` in C++ (it currently just returns `Tensor`), and then extract the correct thing in python. But that would be a breaking change...

Perhaps we could keep both `nonzero` and `nonzero_tuple` in C++, and have them merged into one thing (with a kwarg) in python?

Reply:
> Perhaps we could keep both nonzero and nonzero_tuple in C++, and have them merged into one thing (with a kwarg) in python?

 That sounds good to me.